applications/FluidDynamicsApplication/custom_conditions/navier_stokes_wall_condition.cpp

```diff
@@ -63,6 +65,207 @@ void NavierStokesWallCondition<3,3>::EquationIdVector(EquationIdVectorType& rRes
     }
 }
 
+
+
+template<unsigned int TDim, unsigned int TNumNodes>
+void NavierStokesWallCondition<TDim,TNumNodes>::CalculateLocalSystem(MatrixType& rLeftHandSideMatrix,
+                                      VectorType& rRightHandSideVector,
+                                      ProcessInfo& rCurrentProcessInfo)
+{
+    KRATOS_TRY
+
+    constexpr unsigned int MatrixSize = TNumNodes*(TDim+1);
+    if (rLeftHandSideMatrix.size1() != MatrixSize)
+        rLeftHandSideMatrix.resize(MatrixSize, MatrixSize, false); //false says not to preserve existing storage!!
+    if (rRightHandSideVector.size() != MatrixSize)
+        rRightHandSideVector.resize(MatrixSize, false); //false says not to preserve existing storage!!
+
+    // Struct to pass around the data
+    ConditionDataStruct data;
+    // Allocate memory needed
+    array_1d<double,MatrixSize> rhs_gauss;
+    BoundedMatrix<double,MatrixSize, MatrixSize> lhs_gauss;
+
+    // LHS and RHS contributions initialization
+    noalias(rLeftHandSideMatrix) = ZeroMatrix(MatrixSize,MatrixSize);
+    noalias(rRightHandSideVector) = ZeroVector(MatrixSize);
+    
+    // Compute condition unit normal vector
+    this->CalculateNormal(data.Normal); //this already contains the area
+    const double A = norm_2(data.Normal);
+    data.Normal /= A;
+    
+    // Store the outlet inflow prevention constants in the data structure
+    data.delta = 1e-2; // TODO: Decide if this constant should be fixed or not
+    const ProcessInfo& rProcessInfo = rCurrentProcessInfo; // const to avoid race conditions on data_value_container access/initialization
+    data.charVel = rProcessInfo[CHARACTERISTIC_VELOCITY];
+    
+    // Gauss point information
+    GeometryType& rGeom = this->GetGeometry();
+    const GeometryType::IntegrationPointsArrayType& IntegrationPoints = rGeom.IntegrationPoints(GeometryData::GI_GAUSS_2);
+    const unsigned int NumGauss = IntegrationPoints.size();
+    Vector GaussPtsJDet = ZeroVector(NumGauss);
+    rGeom.DeterminantOfJacobian(GaussPtsJDet, GeometryData::GI_GAUSS_2);
+    const MatrixType Ncontainer = rGeom.ShapeFunctionsValues(GeometryData::GI_GAUSS_2);
+    
+    // Loop on gauss points
+    for(unsigned int igauss = 0; igauss<NumGauss; igauss++)
+    {
+        data.N = row(Ncontainer, igauss);
+        const double J = GaussPtsJDet[igauss];
+        data.wGauss = J * IntegrationPoints[igauss].Weight();
+        ComputeGaussPointRHSContribution(rhs_gauss, data);
+        ComputeGaussPointLHSContribution(lhs_gauss, data);
+        noalias(rLeftHandSideMatrix) += lhs_gauss;
+        noalias(rRightHandSideVector) += rhs_gauss;
+    }
+    
+    // Adding the BEHR2004 contribution if a slip BC is detected
+    // Reference BEHR2004: https://onlinelibrary.wiley.com/doi/abs/10.1002/fld.663
+    if (this->Is(SLIP)){
```

Comment:
All this piece of code should go into the `ComputeGaussPointRHSContribution` and `ComputeGaussPointLHSContribution` methods.

Reply:
I remember that we initially discussed it this way. The problem I was facing, however, is the following:
To use  the function 
`virtual void Calculate(const Variable<array_1d<double, 3 > >& rVariable,
                           array_1d<double, 3 > & Output,
                           const ProcessInfo& rCurrentProcessInfo)`
to retrieve the stress from the element, a reference to a ProcessInfo object is needed. In the given form of the function
 `ComputeGaussPointRHSContribution(array_1d<double,TNumNodes*(TDim+1)>& rhs_gauss,
const ConditionDataStruct& data)`
, this argument is not given. Because ComputeGaussPointRHSContribution(...) is not private, I did not dare and alter it. The only idea I have at the moment would be to retrieve the fluid stress within the function 
`CalculateRightHandSide(VectorType& rRightHandSideVector,
                                    ProcessInfo& rCurrentProcessInfo)`
and pass it via the data struct "data". Do think this to be a viable way? Thank you very much for your opinion in advance.

Reply:
Ok, I hope I found a way around the problem now. Thank you also for the hint concerning the `fluid_element_utilities.h`-include. Many required functions were already there and are now used.

Reply:
Now it looks much better. It makes sense to save the stress in the data structure if the element is flagged as slip.